### ISSUE TITLE
NetworkPolicy ingress podSelector doesn't work for cross-node communication when using native routing

### ISSUE THREAD

**Issue by d-uzlov** [cilium/cilium#32010]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I deployed a network policy that should allow all traffic between pods in the same namespace.
```yaml
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-same-namespace
spec:
  podSelector: {}
  ingress:
  - from:
    - podSelector: {}
```
Some pods are communicating fine, but some are getting timeout as if they weren't matched by `- podSelector: {}`.
I'm able to fix connectivity by removing the pod selector, with a policy like this:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-redis
spec:
  podSelector:
    matchLabels:
      app: redis
  ingress:
  - {}
```
I found out that the "allow-same-namespace" network policy works fine when both pods are on the same node, but if pods are on different nodes pods can't connect.
I then also found out that if I switch from native routing to tunnel mode, network policy starts working as expected across nodes.

### Cilium Version

I tested this with `v1.15.3`, `v1.15.4`, `1.16.0-pre.1`. The issue is present on all of these versions.

### Kernel Version

```
$ uname -a
Linux w1 6.1.0-18-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.76-1 (2024-02-01) x86_64 GNU/Linux
```

### Kubernetes Version

```
$ kubectl version
Client Version: v1.29.3
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.3
```

### Regression

I didn't test this with older versions.

### Sysdump

[cilium-sysdump-20240417-032306.zip](https://github.com/cilium/cilium/files/15002190/cilium-sysdump-20240417-032306.zip)

### Anything else?

<details>
<summary>Helm values</summary>

```yaml
kubeProxyReplacement: true
k8sServicePort: 6443
rollOutCiliumPods: true
operator:
  rollOutPods: true
ipam:
  mode: cluster-pool
  operator:
    clusterPoolIPv4PodCIDRList: 10.201.0.0/16
encryption:
  enabled: false
  type: wireguard

updateStrategy:
  type: RollingUpdate
  rollingUpdate:
    maxUnavailable: 100%

routingMode: native
ipv4NativeRoutingCIDR: 10.3.0.0/16
autoDirectNodeRoutes: true
tunnelProtocol: geneve

envoy:
  enabled: true

hubble:
  tls:
    auto:
      enabled: true
      method: cronJob

l2announcements:
  enabled: false
  leaseDuration: 15s
  leaseRenewDeadline: 5s
  leaseRetryPeriod: 2s

k8sClientRateLimit:
  qps: 100
  burst: 200

pmtuDiscovery:
  enabled: true

externalIPs:
  enabled: true

loadBalancer:
  mode: dsr
  dsrDispatch: opt

bpf:
  masquerade: true

gatewayAPI:
  enabled: true
  secretsNamespace:
    create: false
    name: cilium-secrets
    sync: true
```

</details>

All cluster nodes are in the same L2 segment, there are no firewall rules (or other policies) between nodes.

I see another potentially similar issue but it doesn't mention anything about different nodes and the routing mode, so I'm not sure if it is related to my problem:
- https://github.com/cilium/cilium/issues/30825

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aditighag** [comment]
>Some pods are communicating fine, but some are getting timeout as if they weren't matched by - podSelector: {}.

Can you check Hubble logs to see where the packets might be getting dropped for the affected traffic?

>I then also found out that if I switch from native routing to tunnel mode, network policy starts working as expected across nodes.

This could be the reason why you were seeing connectivity issues for inter node traffic? Did you test connectivity prior to applying the policy?

**Comment by d-uzlov** [comment]
@aditighag thank you for the hints. Using Hubble I found out that this is not a bug but a misconfiguration on my end.
`ipv4NativeRoutingCIDR` was set incorrectly, so pod traffic was being masqueraded, which broke matching.

> This could be the reason why you were seeing connectivity issues for inter node traffic? Did you test connectivity prior to applying the policy?

Yeah, I tested a lot of things. Interestingly, there were no connectivity issues at all, except for pod selector in network policies.

Closing this, sorry to bother you.
